crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -1,11 +1,72 @@
+use alloc::collections::BTreeSet;
+use alloc::vec::Vec;
+
+use miden_lib::note::well_known_note::WellKnownNote;
+use miden_lib::transaction::TransactionKernel;
 use miden_objects::account::AccountId;
 use miden_objects::block::BlockNumber;
+use miden_objects::note::Note;
 use miden_objects::transaction::{InputNote, InputNotes, TransactionArgs};
+use miden_processor::fast::FastProcessor;
 
-use super::{NoteConsumptionInfo, TransactionExecutor};
+use super::TransactionExecutor;
 use crate::auth::TransactionAuthenticator;
 use crate::{DataStore, TransactionExecutorError};
 
+// NOTE CONSUMPTION INFO
+// ================================================================================================
+
+/// Represents a failed note consumption.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct FailedNote {
+    pub note: Note,
+    pub error: TransactionExecutorError,
+}
+
+impl FailedNote {
+    /// Constructs a new `FailedNote`.
+    pub fn new(note: Note, error: TransactionExecutorError) -> Self {
+        Self { note, error }
+    }
+}
+
+/// Contains information about the successful and failed consumption of notes.
+#[derive(Default, Debug)]
+#[non_exhaustive]
+pub struct NoteConsumptionInfo {
+    pub successful: Vec<Note>,
+    pub failed: Vec<FailedNote>,
+}
+
+impl NoteConsumptionInfo {
+    /// Creates a new [`NoteConsumptionInfo`] instance with the given successful notes.
+    pub fn new_successful(successful: Vec<Note>) -> Self {
+        Self { successful, ..Default::default() }
+    }
+
+    /// Creates a new [`NoteConsumptionInfo`] instance with the given successful and failed notes.
+    pub fn new(successful: Vec<Note>, failed: Vec<FailedNote>) -> Self {
+        Self { successful, failed }
+    }
+}
+
+// TRANSACTION EXECUTION ATTEMPT
+// ================================================================================================
+
+/// The result of trying to execute a transaction.
+pub enum TransactionExecutionAttempt {
+    Successful,
+    NoteFailed {
+        failed_note_index: usize,
+        error: TransactionExecutorError,
+    },
+    EpilogueFailed,
+}
```

Comment:
This works, but we could also introduce `NoteConsumptionError` which could look something like:
```rust
pub enum NoteExecutionError {
    TransactionPreparationFailed(TransactionExecutorError),
    PrologueExecutionFailed(TransactionExecutorError),
    EpilogueExecutionFailed(TransactionExecutorError),
    NoteExecutionFailed {
        failed_note_index: usize,
        error: TransactionExecutorError,
    },
}
```
And then `try_execute_notes()` would look something like this:
```rust
async fn try_execute_notes(
    &self,
    account_id: AccountId,
    block_ref: BlockNumber,
    notes: InputNotes<InputNote>,
    tx_args: &TransactionArgs,
) -> Result<(), NoteExecutionError>
```